### ISSUE TITLE
SQLite database concurrency problems and BusyExceptions

### ISSUE THREAD

**Issue by lambi21** [manyfold3d/manyfold#2395]
**Describe the bug**
When starting a scan job on a bigger list of files, some of them didn't generate any previews. When i checked them out individually, some just did not take the jpg in the same folder as preview, others did not show any files at all, as if the scan job did not complete correctly. When I hit "rescan" on one of those models, and then immediately switched to another model an edited it, the error message "We're sorry, but something went wrong." sometimes appeared.
Checking Sidekiq, the log shows a lot of "ActiveRecord::StatementInvalid: SQLite3::BusyException: database is locked" error messages. The same error appeared during the time of the initial scan, which I assume is the reaseon some previews did not get generated (as well as the missing files).
In Sidekiq, I can manually re-run the failed jobs, but only a small fraction of them complete successfully.

**To Reproduce**
1. Have a model folder in manyfold with bigger stl files (not yet scanned by manyfold)
2. hit Rescan
3. immediately switch to another model, edit it and hit save (all while the original scan job is still running)

**Expected behavior**
Correct scanning and editing without missing files or error messages.

**Screenshots**

**Environement*
running as docker on unraid
manyfold-solo 0.72.1

**Comment by Floppy** [comment]
Thanks :+1:

It's iikely these will get automatically retried, but I'll look into ways to reduce tho problem in the first place.

**Comment by lambi21** [comment]
Although the failed jobs will automatically get retried, just looking at it the failure rate seemed a bit high. So I tried again, looking closer at the numbers.
I stopped most of the dockers on the server, so manyfold would have as little interference as possible.
Reset the counter in sidekiq, so all counters showed 0, except for processed which stood at 785.
Then rescanned a new directory, containing 524 files in 105 folders.
This is the result in Sidekiq after ~2mins:
![image](https://github.com/user-attachments/assets/905baed6-fd1d-4d60-80e1-c8f596c44c28)
 (Remember to substract 785 from the "processed" number, since that is where we started).
When I check the "Retries" job category, some of them are on their 4th or 5th retry already.
(Also, in the main manyfold window you get no real feedback if scanning is completed or not, leaving you to check the status in sidekiq)
I let it run, after a while (~10mins) manually starting and restarting the retries when they were scheduled for in 10+ mins, and the counter is now at ~3000 processed vs 850 failed (and 50 still being retried).
I'm no expert on SQLite, but that rate of processed vs failed seems a bit high to me, but I might be wrong.
The problem I see however, since there is no real indication in the web UI when a scan is completed, one might start working on organizing the models when they are not fully scanned. 
Also, it looks like models whose scanning job aborted:
1. do not generate any previews, even on retries (hard to confirm though)
2. do not get assigned any auto-tags (like the !new) tag
Basically what I would have do currently when I have a bigger import is:
1. make sure every model already in the DB has a creator or collection assigned (so I can easily separate any new model coming in)
2. start the import in the evening and let it run overnight, and in the morning check the process window for any remaining retries.

**Comment by lambi21** [comment]
tested some more, I think I narrowed it down: It seems it really is due to too many worker processes trying to access the SQLite DB. I manually changed the worker process number to 1 (using the env variable DEFAULT_WORKER_CONCURRENCY added as separate variable in the unraid docker config window), and re-scanned another folder with 600+ files. Scan took a bit longer, but no failed or retried jobs appeared

**Comment by Floppy** [comment]
I've changed that to the default in v0.72.2 for the solo build. At least then it will work out of the box.

**Comment by Floppy** [comment]
I suspect the answers are in here: https://fractaledmind.github.io/2024/04/15/sqlite-on-rails-the-how-and-why-of-optimal-performance/

**Comment by Floppy** [comment]
I think the changes in the next version will make a huge difference to this!
